### ISSUE TITLE
Consider filtering functionality in `ConstrainedQuadraticModel.iter_constraint_data()`

### ISSUE THREAD

**Issue by arcondello** [dwavesystems/dimod#1381]
Currently [`ConstrainedQuadraticModel.iter_constraint_data()`](https://github.com/dwavesystems/dimod/blob/835688330a6979bb6eeab7d21bbc8f845f37acbe/dimod/constrained/constrained.py#L1147) always iterates over every constraint in the model.

We should consider adding some keyword arguments to make that a bit easier when handling constraints in bulk. I am imagining something like
```Python
     def iter_constraint_data(sample_like, *, filter_function=None):
        if callable(filter_function):
            yield from filter(filter_function, self.iter_constraint_data(sample_like))
            return

        ...
```
and/or
```Python
     def iter_constraint_data(sample_like, *, labels=None):
        ...

        if labels is None:
            labels = self.constraint_labels
            
        for label in labels:
            constraint = self.constraint(label)

            lhs = constraint.lhs.energy((sample, labels))
            rhs = constraint.rhs
            sense = constraint.sense
            ...
```

**Comment by mirko-sof** [comment]
Hi @arcondello,

I am interested in this feature. I have some pre-processing and post-processing steps which start from a feasible solution and then make changes to it by changing the values of a few variables. Every time I make such a change I use the `iter_constraint_data` method to check if the change violates any constraints. When doing this it would save computation time if I could check only those constraints which involve the variables that are about to be changed.

The second approach that you outlined above (passing a `labels` argument) looks to me like it would be suitable for saving computation time. The first approach (passing a `filter_function`) looks like it still iterates over all the constraints because the `filter_function` needs to be applied to each constraint. Is my understanding correct?

**Comment by arcondello** [comment]
> The second approach that you outlined above (passing a labels argument) looks to me like it would be suitable for saving computation time. The first approach (passing a filter_function) looks like it still iterates over all the constraints because the filter_function needs to be applied to each constraint. Is my understanding correct?

Yes exactly. If you know what specific constraints you're looking for by label, we can easily fetch those. But if you want to filter by some other property, or by say a regex of the label, we need to iterate over all constraints.

That said, there may be some performance benefit even in the latter case because you can bypass [the energy calculation](https://github.com/dwavesystems/dimod/blob/835688330a6979bb6eeab7d21bbc8f845f37acbe/dimod/constrained/constrained.py#L1195) which is where most of the time is spent.

If this either/both would be a useful feature for you, we can look into making the change(s) soon.

**Comment by arcondello** [comment]
https://github.com/dwavesystems/dimod/pull/1382 implements the `labels` approach.

**Comment by arcondello** [comment]
@mirko-sof , https://github.com/dwavesystems/dimod/pull/1382 has been merged with the `labels` approach. Are you comfortable testing by installing from source or do you need/want a deploy with the change?

**Comment by mirko-sof** [comment]
I'll try installing from source. Thanks for the quick turnaround.
